x-pack/apm-server/sampling/pubsub/checkpoints.go

```diff
@@ -0,0 +1,86 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package pubsub
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/go-elasticsearch/v7/esapi"
+
+	"github.com/elastic/apm-server/elasticsearch"
+)
+
+// getGlobalCheckpoints returns the current global checkpoint for each index
+// underlying dataStream. Each index is required to have a single (primary) shard.
+func getGlobalCheckpoints(
+	ctx context.Context,
+	client elasticsearch.Client,
+	dataStream string,
+) (map[string]int64, error) {
+	indexGlobalCheckpoints := make(map[string]int64)
+	resp, err := esapi.IndicesStatsRequest{
+		Index: []string{dataStream},
+		Level: "shards",
+		// By default all metrics are returned; query just the "get" metric,
+		// which is very cheap.
+		Metric: []string{"get"},
+	}.Do(ctx, client)
+	if err != nil {
+		return nil, errors.New("index stats request failed")
+	}
+	defer resp.Body.Close()
+	if resp.IsError() {
+		switch resp.StatusCode {
+		case http.StatusNotFound:
+			// Data stream does not yet exist.
+			return indexGlobalCheckpoints, nil
```

Comment:
If I understand that right then for an ES version not supporting the new checkpoint API an empty array would be returned here, but the tail based sampling code would still run, trying to apply sampling, but never receiving checkpoint updates. Have you considered checking the ES version instead and disabling sampling if an ES version < `7.13` is detected?

Reply:
We are not using the new checkpoint API yet. This error case is about the data stream not yet existing, e.g. because no sampled trace IDs have been recorded. In that case we just return an empty map.